Xamarin.Essentials/DeviceInfo/DeviceInfo.uwp.cs

```diff
@@ -29,7 +36,7 @@ static string GetVersionString()
                 var v2 = (v & 0x0000FFFF00000000L) >> 32;
                 var v3 = (v & 0x00000000FFFF0000L) >> 16;
                 var v4 = v & 0x000000000000FFFFL;
-                return $"{v1}.{v2}.{v3}.{v4}";
+                return $"{v1.ToString()}.{v2.ToString()}.{v3.ToString()}.{v4.ToString()}";
```

Comment:
What is the benefit of this change? It is already doing a ToString implicitly.

Reply:
This way you avoid the box. String.format accepts the object type as parameter. If you pass a primitive it will first create a box and then call rostering on the box. This way we avoid the box.

Reply:
This way you avoid the box. String.format accepts the object type as parameter. If you pass a primitive it will first create a box and then call rostering on the box. This way we avoid the box.

Reply:
Interesting. I had a look to see what the IL actually was and it is pretty interesting.

```csharp
int v1 = 123;
Console.WriteLine($"message: {v1} :end");
Console.WriteLine($"message: {v1.ToString()} :end");
```

```csharp
Console.WriteLine(string.Format("message: {0} :end", num));
Console.WriteLine("message: " + num.ToString() + " :end");
```

https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQaAMOaAzAAQDeOJVJaq5l1jAbgIYBOJLJAvCQCQAiFmXQBfFgIDcDRlVYcARj34CFIgHQAVAPYBlYGzAA7AOYAKAJSiFUmSVF3kpZOgBsNACwkAgpfrZZanlOZUFhMWFUcTIicVsA6gcEqicaN08SACE/CmTZYKVeQTV0LT0DY3MrNVQy/UNTS2sYuorG6vjGJKSgA

It goes from boxing, to string allocations. Not sure which is better at this point. Let me do some research... (aka: ask on Twitter 😉)

Reply:
Well the string allocations are obvious, they will incur either way. The string format method will call tostring themselves. If you want some guidance on that, take a look at the heap allocation Roslyn analyzer. They have a rule for this case

Reply:
Here, for further reference the place the actual object.tostring call happens: https://github.com/dotnet/coreclr/blob/a9f3fc16483eecfc47fb79c362811d870be02249/src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs#L1795
This is called from string format here: https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/Common/src/CoreLib/System/String.Manipulation.cs#L528

Reply:
Here, for further reference the place the actual object.tostring call happens: https://github.com/dotnet/coreclr/blob/a9f3fc16483eecfc47fb79c362811d870be02249/src/System.Private.CoreLib/shared/System/Text/StringBuilder.cs#L1795
This is called from string format here: https://github.com/dotnet/corefx/blob/e99ec129cfd594d53f4390bf97d1d736cff6f860/src/Common/src/CoreLib/System/String.Manipulation.cs#L528

Reply:
Okay, after some benchmarking I found out that 4 parameters are a breakpoint (There are overloads for up to three parameters, thus there is no array allocation for the params array). If we were able to guarantee that all numbers being stringified are interned (possible through string.intern) we would be getting a benefit from four parameters onward. I did not know that string.Format has overloads for small parameter packs avoiding the array allocation, the advice about not passing value types to reference type parameters still holds though. In this case I would not use ToString as interning the string definitely brings no noticable benefit, so I reverted the change.